integration-tests/src/tests/client/features/congestion_control.rs

```diff
@@ -84,10 +85,77 @@ fn setup_contract(env: &mut TestEnv) {
     );
 }
 
+/// Check that the congestion info is correctly bootstrapped, updated and
+/// propagated from chunk extra to chunk header. If the
+/// `check_congested_protocol_upgrade` flag is set check that the chain is under
+/// congestion during the protocol upgrade.
+fn check_congestion_info(env: &TestEnv, check_congested_protocol_upgrade: bool) {
```

Comment:
Ah nice! It took me a while to figure out that all the code below looks at the chain state already executed, rather than driving execution. But I really like this approach, since the chain records everything mutably, rather than timing specific events and executing / checking each step, we can just run everything upfront and do all the checks in the end.

I think I could have used this pattern in other integration tests, too, for this and other features. I will put it in my bag of tricks for the future. :D

Reply:
Yeah it worked quite nicely. It initially started as just printing the summary at the end of the test but here I also used it for checking things for the first time. The original inspiration is the absolute log spam horror ;)